DRAFT PR: Feature batch update

be-mevsane opened the pull request.

be-mevsane committed:
Setup and thoughts.

be-mevsane committed:
Update documentation section for 'unclear'.

be-mevsane committed:
Add BatchChangeSet. Add BatchUpdate traits for tuples and impl for tuple of Assign structs. Update first working BatchUpdate test named_struct_batch.

be-mevsane committed:
Merge branch 'diesel-rs:main' into feature_batch_update

weiznich reviewed (commented):
Thanks for opening this draft PR and sorry for taking that long to response. There was just too much stuff going on in the last weeks to take the time to really look into this.

Hopefully the comments I made help to understand how to finish this.

Otherwise I think it might also be a good idea to:

* Add at least one compile test to show that this won't compile with Sqlite/MysqlConnection (it should already be the case, I just want to have a test for this for the future)
* Add a changelog entry as this is a relevant new feature

be-mevsane committed:
update to upstream.

be-mevsane commented on diesel/src/query_builder/update_statement/changeset.rs:
```diff
@@ -140,3 +142,408 @@ where
+            out.push_sql(")");
+        }
+
+        // id = primary key -> therefore not available in update batch! See Note below.
```
> we don't list these keys in the assign in the actual update (that is likely already the case?)

That's true, [AsChangeset](https://github.com/diesel-rs/diesel/blob/main/diesel_derives/src/as_changeset.rs#L48) already handles that

> the update struct contains primary key values. Requiring Identifiable on the struct should give us the right information there. We can list these values as first/last values in the VALUES clause

Nice suggestion. Identifiable provides the values for the primary key and Table::primary_key() privides the column name(s).

On first glance I was unable to find the QueryFragment-like implementations for Identifiable::Id and Table::PrimaryKey. I'll keep searching but hints are welcome if you already know by heart.

be-mevsane committed:
Differentiate between primary keys and regular columns during batch update. Cleanup files. Add theoretical mysql support example.

be-mevsane committed:
Merge branch 'diesel-rs:main' into feature_batch_update

be-mevsane commented:
Hi again @weiznich ,
tldr:
- moved the actual batch update implementations to its correct file.
- Added Backend specific QueryFragment implementations for postgres and mysql.
- postgres works, mysql could work see notes.
- QueryFragment now differentiates between primary key columns and regular columns and both of their associated values.

be-mevsane commented on diesel/src/query_builder/update_statement/batch_update.rs:
```diff
@@ -0,0 +1,436 @@
+//     }
+// }
+#[derive(Debug)]
+pub struct BatchUpdate<I, C, PK, Tab, QId, const STABLE_QUERY_ID: bool> {
```
Similar to BatchInsert I've added `QId` and `STABLE_QUERY_ID` to BatchUpdate. I did not backtrack if that is actually required, but I just did that to stay consistent. One relevant place I found was in `multiconnection` derive.

Should we keep `QId` and `STABLE_QUERY_ID` ?

be-mevsane replied:
Alright, removed.

be-mevsane commented on diesel/src/query_builder/update_statement/batch_update.rs:
```diff
@@ -0,0 +1,436 @@
+        // Return Error by default due to missing support.
+        return Err(QueryBuilderError(Box::new(EmptyChangeset)));
+
+        // TODO:
```
Alternatively, moving `out.push_sql(" SET ");` from [UpdateStatement](https://github.com/diesel-rs/diesel/blob/main/diesel/src/query_builder/update_statement/mod.rs#L206) to its own struct that implements QueryFragment differently for single values or Vectory/Arrays might be another approach.

be-mevsane committed:
Introduce controlled SET clause to UpdateStatement. Update multiconnection snap. Cleanup notes.

be-mevsane committed:
Apply clippy suggestions. Exclude sqlite from tests for now.

weiznich commented on diesel/src/query_builder/update_statement/batch_update.rs:
```diff
@@ -0,0 +1,438 @@
+        //      tab.column_a = tmp.column_a,
+        //      tab.column_c = tmp.column_c;
+
+        // Option 2: mysql 8.x or above
```
It should be fine to go with that variant as that is already implemented. If really someone shows up that needs support for older versions of mysql we still can change the generated SQL. Or rather accept a PR from them to do that. It would be still meaningful for that to keep that comment here.

weiznich reviewed (commented):
Thanks for keeping up the good work. I think we are nearly there. The overall implementation approach looks good to me (beside the handling of the `SET` clause…), there are still a good number of minor style things I would like to see addressed.

Other than that we would like to have some documentation for this, possibly by just having at least one example at `diesel::update` for this behaviour. Finally this is in my opinion a feature that justifies a CHANGELOG entry, so it would be great to have it listed there as well.

be-mevsane commented on diesel/src/query_builder/update_statement/batch_update.rs:
```diff
@@ -0,0 +1,438 @@
+        ambiguous: bool,
+        alias: &'_ str,
+    ) -> QueryResult<()> {
+        match ambiguous {
```
> The values of ambiguous and alias are always the same between calls for the same aschangeset struct at different runtime locations.

Thats holds true.

be-mevsane commented on diesel/src/query_builder/update_statement/batch_update.rs:
```diff
@@ -0,0 +1,438 @@
+}
+
+/// TODO:
+/// Does it make sense to put this to pg/query_builder
```
Sure, done.

be-mevsane commented on diesel/src/query_builder/update_statement/batch_update.rs:
```diff
@@ -0,0 +1,438 @@
+}
+
+/// TODO:
+/// Does it make sense to put this to mysql/query_builder
```
Sure, done.

be-mevsane commented on diesel/src/query_builder/update_statement/changeset.rs:
```diff
@@ -140,3 +150,138 @@ where
+// result in compile error:
+// the parameter type `U` may not live long enough
+// ...so that the reference type `&'a U` does not outlive the data it points at
+impl<'a, U, I, C, PK> AsChangeset for &'a [U]
```
Added.

be-mevsane commented on diesel/src/backend.rs:
```diff
@@ -470,6 +487,39 @@ pub(crate) mod sql_dialect {
+        /// batch update statements to update multiple rows using one
+        /// update statement
+        #[derive(Debug, Copy, Clone)]
+        pub struct PostgresLikeBatchUpdateSupport;
```
Done.

be-mevsane commented on diesel/src/backend.rs:
```diff
@@ -470,6 +487,39 @@ pub(crate) mod sql_dialect {
+        /// batch update statements to update multiple rows using one
+        /// update statement
+        #[derive(Debug, Copy, Clone)]
+        pub struct MySqlLikeBatchUpdateSupport;
```
Done.

be-mevsane commented on diesel/src/backend.rs:
```diff
@@ -470,6 +487,39 @@ pub(crate) mod sql_dialect {
+        #[derive(Debug, Copy, Clone)]
+        pub struct MySqlLikeBatchUpdateSupport;
+
+        impl SupportsBatchUpdate for PostgresLikeBatchUpdateSupport {}
```
Updated all such that `SupportsBatchUpdate` is now required for all that want to support BatchUpdate.

be-mevsane commented on diesel/src/query_builder/update_statement/mod.rs:
```diff
@@ -57,6 +60,7 @@ impl<T: QuerySource, U> UpdateStatement<T, U, SetNotCalled> {
 pub struct UpdateStatement<T: QuerySource, U, V = SetNotCalled, Ret = NoReturningClause> {
     from_clause: T::FromClause,
     where_clause: U,
+    set_clause: SetClause,
```
Assume `struct User { id: i32, name: String, hair_color: String, type: String }` with `id`as the PK:

Calling `AsChangeset::as_changest(&User)` will generate a tuple of three components `(Assign<name, Expr>, Assign<hair_color, Expr>, Assign<type, Expr>)` where the tuple and each Assign implements QueryFragment. There is currently no values "wrapper" other than that tuple generated for regular updates.

Therefore calling `pass.push_sql("SET");` has to be done at the QueryFragment level for the tuple then. Otherwise, that keyword will repeat for each assigned value, in this case 3 times.

That would be here: [QueryFragment implementation ](https://github.com/diesel-rs/diesel/blob/main/diesel/src/type_impls/tuples.rs#L105). This is used globally for other implementations, that expect to just add a comma between the individual elements/fragments.

Since the regular update case does not provide a "values" wrapper similar to the BatchUpdate case I was only able to achieve this through an additional setup.

Note: Adding a values "wrapper" for the regular case is an option here. This would probably introduce more refactoring for the default behavior and for the BatchUpdate as well, since BatchUpdate was implemented by respecting that the `Assign` struct contains individual fields of an update struct rather than all relevant fields for that update struct.

be-mevsane committed:
Merge branch 'diesel-rs:main' into feature_batch_update

be-mevsane committed:
Apply code style. Move impl to respective backend. Cleanup test functions. Add docs to diesel::update.

be-mevsane committed:
Fix stderr line space.

weiznich commented:
I just wanted to leave a note here that I'm aware of the update and I started looking into the changes, but I unfortunately did not finish doing that yet. Hopefully I'm able to do it next week.

weiznich committed:
Merge remote-tracking branch 'upstream/main' into feature_batch_update

weiznich committed:
Cleanup and additional fixes

* Add support for SQLite
* Add support for owned container types (`Vec<_>`, `Box<[_]>`, etc)
* Refactor the internal trait structure to require less traits
* Disallow calling filter or find for the update target
* Add some more tests

weiznich committed:
Merge remote-tracking branch 'upstream/main' into feature_batch_update

weiznich commented:
It took me quite a long time to get back to this. Sorry about that. I finally found the to look into the remaining blockers. As some of them had a quite unclear solution I sat down and did all of the required cleanup from my side.

The commit beside the merges does:

* Add support for sqlite
* Add support for owned container types
* Refactor the internal trait structure to not require that many traits
* Add additional bounds to disallow `filter`/`find` calls for the update statement
* Add more tests

From my perspective this is now ready, but I would appreciate a review from @diesel-rs/reviewers, especially for my additional changes as they are non-trivial.

weiznich committed:
Minor fixes

LucaCappelletti94 commented on diesel/Cargo.toml:
```diff
@@ -71,7 +71,7 @@ tempfile = "3.10.1"
 diesel_test_helper = { path = "../diesel_test_helper" }
 
 [features]
-default = ["with-deprecated", "32-column-tables", "std"]
```
Why are you removing `32-column-tables`?

weiznich replied:
That shouldn't be there. I just tend to remove this feature for local development as it helps with compile times. (Pushed a fix for this)
